### ISSUE TITLE
Weird behavior of Float.to_i

### ISSUE THREAD

**Issue by dabroz** [mruby/mruby#2979]
This code:

    (2**32).to_int

returns a Float on mruby with 32-bit ints. I understand that a regular Ruby would return a Bignum that is missing in (core) mruby. However, this silent float causes many problems (for example when passing as a Random seed). Wouldn't it be better to provide a correct type (which is kind of expected when calling to_int) than a correct value?

**Comment by sagmor** [comment]
Shouldn't it throw an Error?

Returning a silent integer opens the door to even harder to debug problems.

I believe the ideal should be to "try" to convert into a `Bignum` (which should raise something like a `NotImplementedError`). That would allow to implement `Bignum` as a gem for example.

**Comment by matz** [comment]
Due to the absence of Bignums, mruby treats integers and float numbers more uniformly than CRuby.
So when the result of an integer operation is bigger than integer max, it returns a float number.
My rule of thumb is "don't distinguish floats and integers".  They are just numbers.

I admit it is controversial. I am willing to hear your opinion.

**Comment by dabroz** [comment]
I agree that treating both integers and floats simply as numbers works pretty well in Ruby. So it would be nice to have mruby core and bundled gems follow this principle where possible. Currently mruby-random gem doesn't do that for its seeding functions.

However there is still a number of cases where it's just not possible to use floats (like bitwise operators). In that situation it's difficult to correctly handle this silent conversion, especially given that AFAIK there is no built-in way to determine minimum/maximum Fixnum on the Ruby side of mruby (even by using hacks described in [this SO topic](http://stackoverflow.com/questions/535721/ruby-max-integer)). 

There could also be some user-defined C APIs that expect Integers (Fixnums) as arguments to later convert them to machine specific integers. They could try to use `to_int` to provide such typecast. However with the silent integer conversion there is a risk of hard to detect errors that @sagmor mentioned. I did a quick check how mruby behaves compared to CRuby and found a few examples, like:

    # CRuby
    [][2**100] # bignum too big to convert into `long'
    [][2**100 + 0.1] # float 1.2676506e+30 out of range of integer
    # mruby
    [][2**100] # nil
    [][2**100 + 0.1] # nil

However, in this case, the conversion is done manually on the C side (`return (mrb_int)mrb_float(index)`). Maybe `to_int` should diverge from `to_i` and provide the same implementation (or raise an error?) instead of silently failing to typecast?

**Comment by matz** [comment]
Closing — the silent-Float-as-Integer problem you reported has been addressed by several independent changes since 2015:

- mruby's default Integer is now 64-bit on 64-bit hosts, so `(2**32).to_int` returns `Integer` rather than `Float` under the common configuration.

- `Float#to_int` is no longer defined. Calling `.to_int` on a Float now raises `NoMethodError`, which is the "loud failure" @sagmor was asking for instead of a silent type mismatch.

- Out-of-range numeric coercions in places like array indexing now raise `RangeError` instead of silently returning `nil` — the `[][2**100]` and `[][2**100 + 0.1]` cases you flagged are no longer silent.

- The `mruby-bigint` gem provides arbitrary-precision Integer when full Bignum semantics are needed (opt-in via `conf.gem`).

Thanks for the careful original report and the CRuby comparison. The direction the language moved in is essentially what you and @sagmor were asking for.
